File: cli/internal/cmd/library.go
package cmd

import (
	"crypto/sha256"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"syscall"

	"github.com/spf13/cobra"
)

const libraryManifestFilename = ".newsly-agent-manifest.json"

type localLibraryManifest struct {
	Files map[string]string `json:"files"`
}

func (a *App) newLibraryCommand() *cobra.Command {
	libraryCmd := &cobra.Command{
		Use:   "library",
		Short: "Sync the personal markdown library to local disk",
	}

	args := struct {
		Dir           string
		IncludeSource bool
		AllowPruneAll bool
	}{
		IncludeSource: true,
	}

	syncCmd := &cobra.Command{
		Use:   "sync",
		Short: "Download the current markdown library diff to local disk",
		RunE: func(cmd *cobra.Command, _ []string) error {
			runtimeCfg, err := a.resolveRuntimeConfig()
			if err != nil {
				return a.renderError("library.sync", err)
			}
			if err := runtimeCfg.ValidateRemote(); err != nil {
				return a.renderErrorWithPath("library.sync", runtimeCfg.Path, err)
			}

			client, err := a.newRuntimeClient(runtimeCfg)
			if err != nil {
				return a.renderErrorWithPath("library.sync", runtimeCfg.Path, err)
			}

			libraryRoot := runtimeCfg.LibraryRoot
			if strings.TrimSpace(args.Dir) != "" {
				libraryRoot = args.Dir
			}
			if libraryRoot == "" {
				return a.renderError("library.sync", errors.New("missing library root"))
			}
			libraryRoot = filepath.Clean(libraryRoot)
			if err := os.MkdirAll(libraryRoot, 0o700); err != nil {
				return a.renderError("library.sync", err)
			}

			remoteManifest, err := client.GetLibraryManifest(cmd.Context(), args.IncludeSource)
			if err != nil {
				return a.renderErrorWithPath("library.sync", runtimeCfg.Path, err)
			}
			localManifest, err := loadLocalLibraryManifest(
				filepath.Join(libraryRoot, libraryManifestFilename),
			)
			if err != nil {
				return a.renderError("library.sync", err)
			}
			if len(remoteManifest.Documents) == 0 && len(localManifest.Files) > 0 && !args.AllowPruneAll {
				return a.renderError(
					"library.sync",
					errors.New("remote library manifest is empty; refusing to delete all tracked files without --allow-prune-all"),
				)
			}

			downloaded := 0
			unchanged := 0
			repaired := 0
			remoteFiles := make(map[string]string, len(remoteManifest.Documents))
			for _, document := range remoteManifest.Documents {
				remoteFiles[document.RelativePath] = document.ChecksumSHA256
				targetPath, err := safeLibraryPath(libraryRoot, document.RelativePath)
				if err != nil {
					return a.renderError("library.sync", err)
				}
				if localManifest.Files[document.RelativePath] == document.ChecksumSHA256 {
					if actualChecksum, err := checksumFile(targetPath); err == nil && actualChecksum == document.ChecksumSHA256 {
						unchanged++
						continue
					}
					repaired++
				}

				filePayload, err := client.GetLibraryFile(cmd.Context(), document.RelativePath)
				if err != nil {
					return a.renderErrorWithPath("library.sync", runtimeCfg.Path, err)
				}
				if actualChecksum := checksumText(filePayload.Text); actualChecksum != document.ChecksumSHA256 {
					return a.renderError(
						"library.sync",
						fmt.Errorf("downloaded checksum mismatch for %s", document.RelativePath),
					)
				}
				if err := rejectLibrarySymlinks(libraryRoot, targetPath); err != nil {
					return a.renderError("library.sync", err)
				}
				if err := os.MkdirAll(filepath.Dir(targetPath), 0o700); err != nil {
					return a.renderError("library.sync", err)
				}
				if err := writeFileAtomic(targetPath, []byte(filePayload.Text), 0o600); err != nil {
					return a.renderError("library.sync", err)
				}
				downloaded++
			}

			deleted := 0
			for relativePath := range localManifest.Files {
				if _, ok := remoteFiles[relativePath]; ok {
					continue
				}
				targetPath, err := safeLibraryPath(libraryRoot, relativePath)
				if err != nil {
					return a.renderError("library.sync", err)
				}
				if err := os.Remove(targetPath); err != nil && !os.IsNotExist(err) {
					return a.renderError("library.sync", err)
				}
				if err := pruneEmptyLibraryDirs(filepath.Dir(targetPath), libraryRoot); err != nil {
					return a.renderError("library.sync", err)
				}
				deleted++
			}

			if err := saveLocalLibraryManifest(filepath.Join(libraryRoot, libraryManifestFilename), remoteFiles); err != nil {
				return a.renderError("library.sync", err)
			}

			return a.renderSuccess("library.sync", commandResult{
				Data: map[string]any{
					"library_root":   libraryRoot,
					"downloaded":     downloaded,
					"deleted":        deleted,
					"unchanged":      unchanged,
					"repaired":       repaired,
					"document_count": len(remoteManifest.Documents),
				},
			})
		},
	}

	syncCmd.Flags().StringVar(&args.Dir, "dir", "", "Override the local sync directory")
	syncCmd.Flags().BoolVar(&args.IncludeSource, "include-source", true, "Sync source/full-text markdown alongside summaries")
	syncCmd.Flags().BoolVar(&args.AllowPruneAll, "allow-prune-all", false, "Allow an empty remote manifest to delete all tracked local files")

	libraryCmd.AddCommand(syncCmd)
	return libraryCmd
}

func loadLocalLibraryManifest(path string) (localLibraryManifest, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		if os.IsNotExist(err) {
			return localLibraryManifest{Files: map[string]string{}}, nil
		}
		return localLibraryManifest{}, err
	}
	var manifest localLibraryManifest
	if err := json.Unmarshal(data, &manifest); err != nil {
		return localLibraryManifest{}, err
	}
	if manifest.Files == nil {
		manifest.Files = map[string]string{}
	}
	return manifest, nil
}

func saveLocalLibraryManifest(path string, files map[string]string) error {
	manifest := localLibraryManifest{Files: files}
	payload, err := json.MarshalIndent(manifest, "", "  ")
	if err != nil {
		return err
	}
	payload = append(payload, '\n')
	return writeFileAtomic(path, payload, 0o600)
}

func safeLibraryPath(root string, relativePath string) (string, error) {
	cleanRoot := filepath.Clean(root)
	targetPath := filepath.Clean(filepath.Join(cleanRoot, filepath.FromSlash(relativePath)))
	rel, err := filepath.Rel(cleanRoot, targetPath)
	if err != nil {
		return "", err
	}
	if rel == ".." || strings.HasPrefix(rel, ".."+string(filepath.Separator)) {
		return "", errors.New("library path escapes the sync root")
	}
	return targetPath, nil
}

func pruneEmptyLibraryDirs(start string, stop string) error {
	current := filepath.Clean(start)
	stop = filepath.Clean(stop)
	for current != stop && current != "." && current != string(filepath.Separator) {
		if err := os.Remove(current); err != nil {
			if os.IsNotExist(err) {
				current = filepath.Dir(current)
				continue
			}
			if errors.Is(err, syscall.ENOTEMPTY) || errors.Is(err, syscall.EEXIST) {
				return nil
			}
			return err
		}
		current = filepath.Dir(current)
	}
	return nil
}

func checksumFile(path string) (string, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return "", err
	}
	return checksumBytes(data), nil
}

func checksumText(text string) string {
	return checksumBytes([]byte(text))
}

func checksumBytes(data []byte) string {
	sum := sha256.Sum256(data)
	return fmt.Sprintf("%x", sum[:])
}

func writeFileAtomic(path string, data []byte, perm os.FileMode) error {
	tmpPath := path + ".tmp"
	if err := os.WriteFile(tmpPath, data, perm); err != nil {
		return err
	}
	if err := os.Chmod(tmpPath, perm); err != nil {
		_ = os.Remove(tmpPath)
		return err
	}
	if err := os.Rename(tmpPath, path); err != nil {
		_ = os.Remove(tmpPath)
		return err
	}
	return nil
}

func rejectLibrarySymlinks(root string, targetPath string) error {
	root = filepath.Clean(root)
	targetPath = filepath.Clean(targetPath)
	rel, err := filepath.Rel(root, targetPath)
	if err != nil {
		return err
	}
	current := root
	for _, part := range strings.Split(rel, string(filepath.Separator)) {
		if part == "." || part == "" {
			continue
		}
		current = filepath.Join(current, part)
		info, err := os.Lstat(current)
		if err != nil {
			if os.IsNotExist(err) {
				continue
			}
			return err
		}
		if info.Mode()&os.ModeSymlink != 0 {
			return fmt.Errorf("refusing to write through symlink in library path: %s", current)
		}
	}
	return nil
}

File: rust/crates/newsly-cli/src/library.rs
use std::collections::BTreeMap;
use std::fs;
use std::future::Future;
use std::io::Write as _;
use std::path::{Component, Path, PathBuf};

use newsly_contracts::{AgentLibraryFileResponse, AgentLibraryManifestResponse};
use serde::{Deserialize, Serialize};
use sha2::{Digest as _, Sha256};
use tempfile::NamedTempFile;
use thiserror::Error;

pub const LIBRARY_MANIFEST_FILENAME: &str = ".newsbuddy-manifest.json";
pub const LEGACY_LIBRARY_MANIFEST_FILENAME: &str = ".newsly-agent-manifest.json";

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct LibrarySyncReceipt {
    pub library_root: PathBuf,
    pub downloaded: usize,
    pub deleted: usize,
    pub unchanged: usize,
    pub repaired: usize,
    pub document_count: usize,
}

#[derive(Debug, Error)]
pub enum LibrarySyncError<E> {
    #[error("missing library root")]
    MissingRoot,
    #[error(
        "remote library manifest is empty; refusing to delete all tracked files without --allow-prune-all"
    )]
    PruneAllRefused,
    #[error("library path escapes the sync root")]
    EscapingPath,
    #[error("refusing to write through symlink in library path: {0}")]
    Symlink(PathBuf),
    #[error("downloaded checksum mismatch for {0}")]
    ChecksumMismatch(String),
    #[error("library transport failed: {0}")]
    Remote(E),
    #[error(transparent)]
    Io(#[from] std::io::Error),
    #[error(transparent)]
    Json(#[from] serde_json::Error),
    #[error("failed to persist atomic library file: {0}")]
    Persist(String),
}

#[derive(Debug, Default, Deserialize, Serialize)]
struct LocalLibraryManifest {
    #[serde(default)]
    files: BTreeMap<String, String>,
}

/// Apply a remote library manifest to a local directory.
///
/// The caller owns HTTP transport. A file is fetched only when it is new, changed, missing, or
/// locally corrupt, keeping the filesystem safety policy independently testable.
///
/// # Errors
///
/// Returns an error for unsafe paths or symlinks, invalid manifests or checksums, local filesystem
/// failures, remote fetch failures, or an unapproved full prune.
pub async fn sync_library<F, Fut, E>(
    root: &Path,
    remote: &AgentLibraryManifestResponse,
    allow_prune_all: bool,
    mut fetch_file: F,
) -> Result<LibrarySyncReceipt, LibrarySyncError<E>>
where
    F: FnMut(String) -> Fut,
    Fut: Future<Output = Result<AgentLibraryFileResponse, E>>,
{
    if root.as_os_str().is_empty() {
        return Err(LibrarySyncError::MissingRoot);
    }
    fs::create_dir_all(root)?;
    set_directory_mode(root)?;

    let canonical_manifest_path = root.join(LIBRARY_MANIFEST_FILENAME);
    let legacy_manifest_path = root.join(LEGACY_LIBRARY_MANIFEST_FILENAME);
    let local = load_local_manifest(&canonical_manifest_path, &legacy_manifest_path)?;
    if remote.documents.is_empty() && !local.files.is_empty() && !allow_prune_all {
        return Err(LibrarySyncError::PruneAllRefused);
    }

    let mut downloaded = 0;
    let mut unchanged = 0;
    let mut repaired = 0;
    let mut desired = BTreeMap::new();

    for document in &remote.documents {
        desired.insert(
            document.relative_path.clone(),
            document.checksum_sha256.clone(),
        );
        let target = safe_library_path(root, &document.relative_path)?;
        if local.files.get(&document.relative_path) == Some(&document.checksum_sha256) {
            if matches!(
                checksum_file(&target),
                Ok(actual) if actual == document.checksum_sha256
            ) {
                unchanged += 1;
                continue;
            }
            repaired += 1;
        }

        let payload = fetch_file(document.relative_path.clone())
            .await
            .map_err(LibrarySyncError::Remote)?;
        if checksum_bytes(payload.text.as_bytes()) != document.checksum_sha256 {
            return Err(LibrarySyncError::ChecksumMismatch(
                document.relative_path.clone(),
            ));
        }
        reject_library_symlinks(root, &target)?;
        let parent = target.parent().ok_or(LibrarySyncError::EscapingPath)?;
        fs::create_dir_all(parent)?;
        set_directory_tree_mode(root, parent)?;
        write_file_atomic(&target, payload.text.as_bytes(), 0o600)?;
        downloaded += 1;
    }

    let mut deleted = 0;
    for relative_path in local.files.keys() {
        if desired.contains_key(relative_path) {
            continue;
        }
        let target = safe_library_path(root, relative_path)?;
        reject_library_symlinks(root, &target)?;
        match fs::remove_file(&target) {
            Ok(()) => {
                if let Some(parent) = target.parent() {
                    prune_empty_library_dirs(parent, root)?;
                }
            }
            Err(error) if error.kind() == std::io::ErrorKind::NotFound => {}
            Err(error) => return Err(error.into()),
        }
        deleted += 1;
    }

    save_local_manifest(&canonical_manifest_path, &desired)?;
    Ok(LibrarySyncReceipt {
        library_root: root.to_path_buf(),
        downloaded,
        deleted,
        unchanged,
        repaired,
        document_count: remote.documents.len(),
    })
}

fn load_local_manifest(
    canonical_path: &Path,
    legacy_path: &Path,
) -> Result<LocalLibraryManifest, std::io::Error> {
    let data = match fs::read(canonical_path) {
        Ok(data) => data,
        Err(error) if error.kind() == std::io::ErrorKind::NotFound => match fs::read(legacy_path) {
            Ok(data) => data,
            Err(error) if error.kind() == std::io::ErrorKind::NotFound => {
                return Ok(LocalLibraryManifest::default());
            }
            Err(error) => return Err(error),
        },
        Err(error) => return Err(error),
    };
    serde_json::from_slice(&data).map_err(std::io::Error::other)
}

fn save_local_manifest<E>(
    path: &Path,
    files: &BTreeMap<String, String>,
) -> Result<(), LibrarySyncError<E>> {
    let mut payload = serde_json::to_vec_pretty(&LocalLibraryManifest {
        files: files.clone(),
    })?;
    payload.push(b'\n');
    write_file_atomic(path, &payload, 0o600)
}

fn safe_library_path<E>(root: &Path, relative_path: &str) -> Result<PathBuf, LibrarySyncError<E>> {
    let relative = Path::new(relative_path);
    if relative.is_absolute()
        || relative.components().any(|component| {
            matches!(
                component,
                Component::ParentDir | Component::RootDir | Component::Prefix(_)
            )
        })
    {
        return Err(LibrarySyncError::EscapingPath);
    }
    Ok(root.join(relative))
}

fn reject_library_symlinks<E>(root: &Path, target: &Path) -> Result<(), LibrarySyncError<E>> {
    let relative = target
        .strip_prefix(root)
        .map_err(|_| LibrarySyncError::EscapingPath)?;
    let mut current = root.to_path_buf();
    for component in relative.components() {
        let Component::Normal(part) = component else {
            continue;
        };
        current.push(part);
        match fs::symlink_metadata(&current) {
            Ok(metadata) if metadata.file_type().is_symlink() => {
                return Err(LibrarySyncError::Symlink(current));
            }
            Ok(_) => {}
            Err(error) if error.kind() == std::io::ErrorKind::NotFound => {}
            Err(error) => return Err(error.into()),
        }
    }
    Ok(())
}

fn checksum_file(path: &Path) -> Result<String, std::io::Error> {
    fs::read(path).map(|bytes| checksum_bytes(&bytes))
}

fn checksum_bytes(bytes: &[u8]) -> String {
    use std::fmt::Write as _;

    Sha256::digest(bytes)
        .iter()
        .fold(String::with_capacity(64), |mut output, byte| {
            let _ = write!(output, "{byte:02x}");
            output
        })
}

fn write_file_atomic<E>(path: &Path, data: &[u8], mode: u32) -> Result<(), LibrarySyncError<E>> {
    let parent = path.parent().ok_or(LibrarySyncError::EscapingPath)?;
    fs::create_dir_all(parent)?;
    let mut temporary = NamedTempFile::new_in(parent)?;
    temporary.write_all(data)?;
    temporary.as_file().sync_all()?;
    set_file_mode(temporary.path(), mode)?;
    temporary
        .persist(path)
        .map_err(|error| LibrarySyncError::Persist(error.error.to_string()))?;
    Ok(())
}

fn prune_empty_library_dirs<E>(start: &Path, stop: &Path) -> Result<(), LibrarySyncError<E>> {
    let mut current = start.to_path_buf();
    while current != stop {
        match fs::remove_dir(&current) {
            Ok(()) => {}
            Err(error) if error.kind() == std::io::ErrorKind::NotFound => {}
            Err(error) if error.kind() == std::io::ErrorKind::DirectoryNotEmpty => return Ok(()),
            Err(error) => return Err(error.into()),
        }
        let Some(parent) = current.parent() else {
            break;
        };
        current = parent.to_path_buf();
    }
    Ok(())
}

fn set_directory_tree_mode<E>(root: &Path, leaf: &Path) -> Result<(), LibrarySyncError<E>> {
    let relative = leaf
        .strip_prefix(root)
        .map_err(|_| LibrarySyncError::EscapingPath)?;
    let mut current = root.to_path_buf();
    set_directory_mode(&current)?;
    for component in relative.components() {
        if let Component::Normal(part) = component {
            current.push(part);
            set_directory_mode(&current)?;
        }
    }
    Ok(())
}

#[cfg(unix)]
fn set_directory_mode(path: &Path) -> Result<(), std::io::Error> {
    use std::os::unix::fs::PermissionsExt as _;
    fs::set_permissions(path, fs::Permissions::from_mode(0o700))
}

#[cfg(not(unix))]
fn set_directory_mode(_path: &Path) -> Result<(), std::io::Error> {
    Ok(())
}

#[cfg(unix)]
fn set_file_mode(path: &Path, mode: u32) -> Result<(), std::io::Error> {
    use std::os::unix::fs::PermissionsExt as _;
    fs::set_permissions(path, fs::Permissions::from_mode(mode))
}

#[cfg(not(unix))]
fn set_file_mode(_path: &Path, _mode: u32) -> Result<(), std::io::Error> {
    Ok(())
}

#[cfg(test)]
mod tests {
    use std::convert::Infallible;

    use serde_json::json;
    use tempfile::tempdir;

    use super::*;

    fn remote_manifest(path: Option<&str>, text: &str) -> AgentLibraryManifestResponse {
        let documents = path.map_or_else(Vec::new, |path| {
            vec![json!({
                "relative_path": path,
                "content_id": 1,
                "variant": "summary",
                "updated_at": "2026-08-31T12:00:00Z",
                "size_bytes": text.len(),
                "checksum_sha256": checksum_bytes(text.as_bytes())
            })]
        });
        serde_json::from_value(json!({
            "generated_at": "2026-08-31T12:00:00Z",
            "include_source": true,
            "documents": documents
        }))
        .unwrap()
    }

    fn file_payload(path: &str, text: &str) -> AgentLibraryFileResponse {
        serde_json::from_value(json!({
            "relative_path": path,
            "content_id": 1,
            "variant": "summary",
            "updated_at": "2026-08-31T12:00:00Z",
            "checksum_sha256": checksum_bytes(text.as_bytes()),
            "text": text
        }))
        .unwrap()
    }

    #[tokio::test]
    async fn downloads_then_leaves_a_verified_file_unchanged() {
        let directory = tempdir().unwrap();
        let path = "article/example.md";
        let text = "# Example\n";
        let remote = remote_manifest(Some(path), text);
        let first = sync_library(directory.path(), &remote, false, |requested| async move {
            Ok::<_, Infallible>(file_payload(&requested, text))
        })
        .await
        .unwrap();
        assert_eq!(first.downloaded, 1);
        assert_eq!(
            fs::read_to_string(directory.path().join(path)).unwrap(),
            text
        );

        let second = sync_library(directory.path(), &remote, false, |_| async {
            panic!("unchanged file must not be fetched");
            #[allow(unreachable_code)]
            Ok::<AgentLibraryFileResponse, Infallible>(file_payload(path, text))
        })
        .await
        .unwrap();
        assert_eq!(second.unchanged, 1);
        assert_eq!(second.downloaded, 0);
    }

    #[tokio::test]
    async fn repairs_a_corrupt_tracked_file() {
        let directory = tempdir().unwrap();
        let path = "article/example.md";
        let text = "# Correct\n";
        let remote = remote_manifest(Some(path), text);
        sync_library(directory.path(), &remote, false, |requested| async move {
            Ok::<_, Infallible>(file_payload(&requested, text))
        })
        .await
        .unwrap();
        fs::write(directory.path().join(path), "corrupt").unwrap();

        let receipt = sync_library(directory.path(), &remote, false, |requested| async move {
            Ok::<_, Infallible>(file_payload(&requested, text))
        })
        .await
        .unwrap();
        assert_eq!(receipt.repaired, 1);
        assert_eq!(receipt.downloaded, 1);
    }

    #[tokio::test]
    async fn empty_remote_requires_explicit_prune_permission() {
        let directory = tempdir().unwrap();
        let path = "article/example.md";
        fs::create_dir_all(directory.path().join("article")).unwrap();
        fs::write(directory.path().join(path), "old").unwrap();
        fs::write(
            directory.path().join(LIBRARY_MANIFEST_FILENAME),
            format!("{{\"files\":{{\"{path}\":\"old\"}}}}"),
        )
        .unwrap();
        let error = sync_library(
            directory.path(),
            &remote_manifest(None, ""),
            false,
            |_| async {
                panic!("empty manifest has no downloads");
                #[allow(unreachable_code)]
                Ok::<AgentLibraryFileResponse, Infallible>(file_payload(path, ""))
            },
        )
        .await
        .unwrap_err();
        assert!(matches!(error, LibrarySyncError::PruneAllRefused));
        assert!(directory.path().join(path).exists());
    }

    #[tokio::test]
    async fn rejects_escaping_paths_before_fetching() {
        let directory = tempdir().unwrap();
        let remote = remote_manifest(Some("../escape.md"), "bad");
        let error = sync_library(directory.path(), &remote, false, |_| async {
            panic!("unsafe paths must not be fetched");
            #[allow(unreachable_code)]
            Ok::<AgentLibraryFileResponse, Infallible>(file_payload("unused", ""))
        })
        .await
        .unwrap_err();
        assert!(matches!(error, LibrarySyncError::EscapingPath));
    }

    #[tokio::test]
    async fn rejects_a_downloaded_checksum_mismatch() {
        let directory = tempdir().unwrap();
        let path = "article/example.md";
        let remote = remote_manifest(Some(path), "expected");
        let error = sync_library(directory.path(), &remote, false, |requested| async move {
            Ok::<_, Infallible>(file_payload(&requested, "different"))
        })
        .await
        .unwrap_err();
        assert!(matches!(error, LibrarySyncError::ChecksumMismatch(value) if value == path));
        assert!(!directory.path().join(path).exists());
    }

    #[cfg(unix)]
    #[tokio::test]
    async fn rejects_writes_through_symlinks() {
        use std::os::unix::fs::symlink;

        let directory = tempdir().unwrap();
        let outside = tempdir().unwrap();
        symlink(outside.path(), directory.path().join("article")).unwrap();
        let path = "article/example.md";
        let remote = remote_manifest(Some(path), "bad");
        let error = sync_library(directory.path(), &remote, false, |requested| async move {
            Ok::<_, Infallible>(file_payload(&requested, "bad"))
        })
        .await
        .unwrap_err();
        assert!(matches!(error, LibrarySyncError::Symlink(_)));
        assert!(!outside.path().join("example.md").exists());
    }
}
